File: src/p2_add_two_numbers.py
#
# @lc app=leetcode id=2 lang=python3
#
# [2] Add Two Numbers
#

from structs import ListNode


# @lc code=start
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        head = ListNode(0)
        p = head
        carry = 0

        while p != None:
            if l1 != None:
                p.val += l1.val
                l1 = l1.next

            if l2 != None:
                p.val += l2.val
                l2 = l2.next

            if p.val >= 10:
                p.val -= 10
                carry = 1
            else:
                carry = 0

            if l1 != None or l2 != None or carry != 0:
                p.next = ListNode(carry)

            p = p.next

        return head


# @lc code=end

if __name__ == "__main__":
    s = Solution()
    l1 = ListNode(2)**ListNode(4)**ListNode(3)
    l2 = ListNode(5)**ListNode(6)**ListNode(4)

    print(s.addTwoNumbers(l1, l2))

File: src/p2.addTwoNumbers.c
/*
 * @lc app=leetcode id=2 lang=c
 *
 * [2] Add Two Numbers
 */

#include "structs.h"

// @lc code=start
typedef struct ListNode ListNode;

struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    ListNode* head = (ListNode*)malloc(sizeof(ListNode));

    ListNode* p = head;
    for (int carry = 0; p != NULL; p = p->next) {
        p->val = carry;

        if (l1 != NULL) {
            p->val += l1->val;
            l1 = l1->next;
        }
        if (l2 != NULL) {
            p->val += l2->val;
            l2 = l2->next;
        }

        if (p->val >= 10) {
            p->val -= 10;
            carry = 1;
        } else {
            carry = 0;
        }

        p->next = (l1 != NULL || l2 != NULL || carry != 0) ? (ListNode*)malloc(sizeof(ListNode)) : NULL;
    }

    return head;
}
// @lc code=end

int main(int argc, char const* argv[]) {
    ListNode* l1 = arrayToListNode((int[]){2, 4, 3}, 3);
    ListNode* l2 = arrayToListNode((int[]){5, 6, 4}, 3);
    ListNode* l3 = arrayToListNode((int[]){7, 0, 8}, 3);

    ListNode* actual = addTwoNumbers(l1, l2);

    printf("expected: \t");
    printListNode(l3);
    printf("\nactual: \t");
    printListNode(actual);
    puts("");

    freeListNode(l1);
    freeListNode(l2);
    freeListNode(l3);
    freeListNode(actual);

    return 0;
}
